janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -17,81 +17,236 @@
+                } else {
+                    subLimit = baseSubLimit;
+                }
+                subLimit = Math.min(Query.NO_LIMIT, subLimit);
+
+                executeQueries(queries, idx, subLimit, offsets, subResultToQueryMap, indexSerializer, tx);
+
+                if (offsets[idx] < subLimit) {
+                    resultsExhausted[idx] = true;
+                    resultsExhaustedCount++;
+                }
+            }
+
+            intersectResults(subResultToQueryMap, results, limit, queries.size());
+
+            scores = computeScores(resultsExhaustedCount, resultsExhausted, offsets, queries.size(), results.size(),
+                limit, subResultToQueryMap);
+
+        } while (resultsExhaustedCount < queries.size() && results.size() < limit);
+        elementIterator = results.stream().limit(limit).map(conversionFunction).map(r -> (JanusGraphElement) r).iterator();
+    }
+
+    private void executeQueries(List<JointIndexQuery.Subquery> queries, final int idx, final int subLimit,
+                                int[] offsets, Map<Object, List<Integer>> subResultToQueryMap,
+                                IndexSerializer indexSerializer, BackendTransaction tx) {
+        JointIndexQuery.Subquery subQuery = queries.get(idx).updateLimit(subLimit);
+        final List<Object> subQueryCache = indexCache.getIfPresent(subQuery);
+        if (subQueryCache != null) {
+            assert subQueryCache.size() >= offsets[idx];
+            for (int j = offsets[idx]; j < subQueryCache.size(); j++) {
+                subResultToQueryMap.computeIfAbsent(subQueryCache.get(j), k -> new ArrayList<>()).add(idx);
+            }
+            offsets[idx] = subQueryCache.size();
+        } else {
+            QueryProfiler profiler = subQuery.getProfiler();
+            QueryProfiler.startProfile(profiler, subQuery);
+            // TODO: leverage the scrolling capability of indexing backends ratherthan throw away old results
+            indexSerializer.query(subQuery, tx).skip(offsets[idx]).forEachOrdered(result -> {
+                offsets[idx]++;
+                subResultToQueryMap.computeIfAbsent(result, k -> new ArrayList<>()).add(idx);
+            });
             profiler.stopTimer();
+            profiler.setResultSize(offsets[idx]);
+            // TODO: should we put it into cache?
+        }
+
+    }
+
+    private void intersectResults(Map<Object, List<Integer>> subResultToQueryMap,
+                                  List<Object> results, final int limit, final int numberOfQueries) {
+        for (Iterator<Map.Entry<Object, List<Integer>>> it = subResultToQueryMap.entrySet().iterator();
+             it.hasNext(); ) {
+            Map.Entry<Object, List<Integer>> entry = it.next();
+            if (entry.getValue().size() == numberOfQueries) {
+                // this particular result satisfies every index query
+                results.add(entry.getKey());
```

Comment:
How do you ensure an entry gets only added to the result list once here? `intersectResults` is called in every iteration of the outer loop in `intersectAll`. But in the (N+1)st iteration, `subResultToQueryMap` still contains all the valid results that were found in the N-th iteration or am I missing something?

Reply:
On [line 220](https://github.com/JanusGraph/janusgraph/pull/2040/files/7de65fb3da9fec3b9c05ab9c18c6b91dd55d0204#diff-d3269581786b8b3f748ed9ec0b653afbR220) I removed it from `subResultToQueryMap`

Reply:
I should have looked just two lines further... :sweat_smile: 
Resolved!